File: Python/PR/BasicGraph.py
import time

class BasicGraph:
    def __init__(self):
        self.adjList = dict()

    def add_edge(self, src, dest):
        if src not in self.adjList:
            self.adjList[src] = AdjList()
        self.adjList[src].out += 1
        if dest not in self.adjList:
            self.adjList[dest] = AdjList()
        self.adjList[dest].aList.append(src)

    def calculate_page_rank(self, iterations):
        d = 0.85
        n = len(self.adjList)
        sink_nodes = []
        npg = dict()
        for node in self.adjList:
            self.adjList[node].pr = 1.0 / n
            if self.adjList[node].out == 0:
                sink_nodes.append(node)
        num = iterations
        while iterations > 0:
            start = time.time()
            sink_node_rank = 0
            for node in sink_nodes:
                sink_node_rank += self.adjList[node].pr
            for node in self.adjList:
                npg[node] = (1.0 - d) / n
                npg[node] += (d * sink_node_rank) / n
                for inNode in self.adjList[node].aList:
                    npg[node] += ((d * self.adjList[inNode].pr) / self.adjList[inNode].out)
            for node in self.adjList:
                self.adjList[node].pr = npg[node]
            iterations -= 1
            end = time.time() - start
            print("Iteration " + str(num - iterations)) + " completed in " + str(end * 1000) + "ms"

    def print_graph(self):
        for node in self.adjList:
            print("Inverted AdjList for Vertex " + node)
            print("Out: " + str(self.adjList[node].out))
            for inNode in self.adjList[node].aList:
                print " <-" + inNode,
            print('')

    def print_scores(self):
        total = 0
        for node in self.adjList:
            print(node + "\t" + str(self.adjList[node].pr))
            total += self.adjList[node].pr
        print("Sum: " + str(total))


class AdjList:
    def __init__(self):
        self.pr = 0
        self.out = 0
        self.aList = []

File: Swift/PageRank/BasicGraph.swift
//
//  BasicGraph.swift
//  PageRank
//
//  Created by 奥玛 on 4/14/16.
//  Copyright © 2016 THU. All rights reserved.
//

import Foundation

struct AdjList {
    var pr: Float = 0.0
    var out: Int = 0
    var inList: Array = [Int]()
}

class BasicGraph {
    var map: [Int: AdjList] = [:]
    
    func addEdge(src: Int, dest: Int) -> Void {
        if(map[src]==nil){
            map[src] = AdjList()
        }
        map[src]?.out += 1
        if(map[dest]==nil){
            map[dest] = AdjList()
        }
        map[dest]?.inList.append(src)
    }
    
    func calculatePageRank(iterations: Int) -> Void {
        let d: Float = 0.85
        let N: Float = Float(map.count)
        let restartRate = (1.0 - d) / N
        var sinkNodes: Array = [Int]()
        var newPageRank: [Int: Float] = [:]
        for (key, node) in map {
            map[key]?.pr = 1.0 / N
            if(node.out==0){
                sinkNodes.append(key)
            }
        }
        var num = iterations
        while num > 0 {
            let start = NSDate()
            var sinkNodeRank: Float = 0
            for node in sinkNodes {
                sinkNodeRank += (map[node]?.pr)!
            }
            for (key, node) in map {
                var npg: Float = restartRate
                npg += (d * sinkNodeRank) / N
                for inNode in node.inList {
                    npg += ((d * map[inNode]!.pr) / Float(map[inNode]!.out))
                }
                newPageRank[key] = npg
            }
            for (key, _) in map {
                map[key]!.pr = newPageRank[key]!
            }
            num -= 1
            let end = NSDate()
            let timeInterval: Double = end.timeIntervalSinceDate(start)
            print("Iteration \(iterations - num) completed in \(timeInterval * 1000)")
        }
    }
    
    func printScores() -> Void {
        for (key, node) in map {
            print("\(key)\t\(node.pr)")
        }
    }
    
    func printGraph() -> Void {
        for (key, node) in map {
            print("Inverted AdjList for vertex \(key) \nOut: \(node.out)")
            for inNode in node.inList {
                print("<- \(inNode) ", terminator:"")
            }
            print("")
        }
    }
}